cni/pkg/plugin/plugin_dryrun_test.go

```diff
@@ -0,0 +1,195 @@
+				Annotations: map[string]string{
+					annotation.SidecarStatus.Name:                         "true",
+					annotation.SidecarTrafficIncludeOutboundIPRanges.Name: "127.0.0.0/8",
+					annotation.SidecarTrafficExcludeOutboundIPRanges.Name: "10.0.0.0/8",
+				},
+				ProxyEnvironments: map[string]string{},
+			},
+			golden: filepath.Join(env.IstioSrc, "cni/pkg/plugin/testdata/include-exclude-ip.txt.golden"),
+		},
+		{
+			name: "include-exclude-ports",
+			input: &PodInfo{
+				Containers:     []string{"test", "istio-proxy"},
+				InitContainers: map[string]struct{}{"istio-validate": {}},
+				Annotations: map[string]string{
+					annotation.SidecarStatus.Name:                      "true",
+					annotation.SidecarTrafficIncludeInboundPorts.Name:  "1111,2222",
+					annotation.SidecarTrafficExcludeInboundPorts.Name:  "3333,4444",
+					annotation.SidecarTrafficExcludeOutboundPorts.Name: "5555,6666",
+				},
+				ProxyEnvironments: map[string]string{},
+			},
+			golden: filepath.Join(env.IstioSrc, "cni/pkg/plugin/testdata/include-exclude-ports.txt.golden"),
+		},
+		{
+			name: "tproxy",
+			input: &PodInfo{
+				Containers:     []string{"test", "istio-proxy"},
+				InitContainers: map[string]struct{}{"istio-validate": {}},
+				Annotations: map[string]string{
+					annotation.SidecarStatus.Name:           "true",
+					annotation.SidecarInterceptionMode.Name: redirectModeTPROXY,
+				},
+				ProxyEnvironments: map[string]string{},
+			},
+			golden: filepath.Join(env.IstioSrc, "cni/pkg/plugin/testdata/tproxy.txt.golden"),
+		},
+		{
+			name: "DNS",
+			input: &PodInfo{
+				Containers:        []string{"test", "istio-proxy"},
+				InitContainers:    map[string]struct{}{"istio-validate": {}},
+				Annotations:       map[string]string{annotation.SidecarStatus.Name: "true"},
+				ProxyEnvironments: map[string]string{options.DNSCaptureByAgent.Name: "true"},
+			},
+			golden: filepath.Join(env.IstioSrc, "cni/pkg/plugin/testdata/dns.txt.golden"),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// TODO(bianpengyuan): How do we test ipv6 rules?
+			getKubePodInfo = generateMockK8sPodInfoFunc(tt.input)
+			tmpDir := t.TempDir()
+			outputFilePath := filepath.Join(tmpDir, "output.txt")
+			if _, err := os.Create(outputFilePath); err != nil {
+				t.Fatalf("Failed to create temp file for IPTables rule output: %v", err)
+			}
+			os.Setenv(dryRunFilePath.Name, outputFilePath)
+			_, _, err := testutils.CmdAddWithResult(
+				sandboxDirectory, ifname, []byte(cniConf), func() error { return CmdAdd(args) })
+			os.Unsetenv(dryRunFilePath.Name)
+			if err != nil {
+				t.Fatalf("CNI cmdAdd failed with error: %v", err)
+			}
+
+			generated, err := ioutil.ReadFile(outputFilePath)
+			if err != nil {
+				log.Fatalf("Cannot read generated IPTables rule file: %v", err)
+			}
+			generatedRules := getRules(generated)
+
+			if os.Getenv("REFRESH_GOLDEN") == "true" {
```

Comment:
pilot/test/util/diff.go has some useful functions for golden files, including CompareContent, RefreshGoldenFile, etc that may be preferred. Not sure how well it would work here with the custom logic, but it seems like it could be plausible

Reply:
Thanks for the pointer! I used the RefreshGolden function, Compare function is a bit tricky because IPTables output tables order is not stable so need a bit of custom logic. IPTables output is small and I think we can live with raw got/want output.